File: docs-design/examples/ralf/ralf_subscriber.py
#!/usr/bin/env python3
"""Full working RALF subscriber example.

Uses ralf_parser.py to parse incoming gateway messages.
"""

from __future__ import annotations

import argparse
import socket
from typing import Iterable

from ralf_parser import build_ralf_line, parse_ralf_line


class LineReader:
    def __init__(self, sock: socket.socket) -> None:
        self.sock = sock
        self.buf = bytearray()

    def recv_line(self) -> str:
        while True:
            nl = self.buf.find(b"\n")
            if nl >= 0:
                line = bytes(self.buf[:nl])
                del self.buf[: nl + 1]
                return line.decode("utf-8", errors="replace")

            chunk = self.sock.recv(4096)
            if not chunk:
                raise RuntimeError("gateway closed connection")
            self.buf.extend(chunk)


def send_line(sock: socket.socket, msg_type: str, fields: dict[str, str]) -> None:
    sock.sendall(build_ralf_line(msg_type, fields).encode("utf-8"))


def subscribe_channels(sock: socket.socket, role: str, channels: Iterable[str], symbols: str) -> None:
    for ch in channels:
        send_line(sock, "SUB", {"ROLE": role, "CH": ch, "SYM": symbols})


def main() -> None:
    parser = argparse.ArgumentParser(description="RALF subscriber example")
    parser.add_argument("--host", default="127.0.0.1")
    parser.add_argument("--port", type=int, default=5580)
    parser.add_argument("--client", default="ext-client-01")
    parser.add_argument("--role", default="CLEARING", choices=["CLEARING", "DROP_COPY", "AUDIT"])
    parser.add_argument("--lastseq", default="0")
    parser.add_argument(
        "--channels",
        default="CLEARING,DROP_COPY,AUDIT",
        help="Comma-separated channels to subscribe",
    )
    parser.add_argument("--symbols", default="*", help="Symbol scope for SUB")
    args = parser.parse_args()

    channels = [c.strip() for c in args.channels.split(",") if c.strip()]

    with socket.create_connection((args.host, args.port), timeout=5) as sock:
        reader = LineReader(sock)
        send_line(
            sock,
            "HELLO",
            {
                "CLIENT": args.client,
                "PROTO": "RALF1",
                "ROLE": args.role,
                "LASTSEQ": args.lastseq,
            },
        )

        welcome = parse_ralf_line(reader.recv_line())
        print(f"WELCOME: type={welcome.msg_type} fields={welcome.fields}")

        subscribe_channels(sock, args.role, channels, args.symbols)
        print(f"Subscribed channels={channels} as role={args.role}")

        while True:
            msg = parse_ralf_line(reader.recv_line())
            print(f"{msg.msg_type} {msg.fields}")


if __name__ == "__main__":
    main()

File: docs-design/examples/ralf/ralf_subscriber.c
#include "ralf_parser.h"

#include <arpa/inet.h>
#include <errno.h>
#include <netinet/in.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/socket.h>
#include <sys/types.h>
#include <unistd.h>

static int send_line(int fd, const char *line) {
    size_t len = strlen(line);
    if (write(fd, line, len) != (ssize_t)len) return -1;
    if (write(fd, "\n", 1) != 1) return -1;
    return 0;
}

static int recv_line(int fd, char *buf, size_t cap) {
    size_t n = 0;
    while (n + 1 < cap) {
        char c;
        ssize_t r = read(fd, &c, 1);
        if (r <= 0) return -1;
        if (c == '\n') break;
        buf[n++] = c;
    }
    buf[n] = '\0';
    return 0;
}

static int connect_gateway(const char *host, int port) {
    int fd = socket(AF_INET, SOCK_STREAM, 0);
    if (fd < 0) return -1;

    struct sockaddr_in addr;
    memset(&addr, 0, sizeof(addr));
    addr.sin_family = AF_INET;
    addr.sin_port = htons((uint16_t)port);

    if (inet_pton(AF_INET, host, &addr.sin_addr) != 1) {
        close(fd);
        return -1;
    }

    if (connect(fd, (struct sockaddr *)&addr, sizeof(addr)) != 0) {
        close(fd);
        return -1;
    }

    return fd;
}

int main(int argc, char **argv) {
    const char *host = "127.0.0.1";
    int port = 5580;
    const char *role = "CLEARING";

    if (argc > 1) host = argv[1];
    if (argc > 2) port = atoi(argv[2]);
    if (argc > 3) role = argv[3];

    int fd = connect_gateway(host, port);
    if (fd < 0) {
        perror("connect_gateway");
        return 1;
    }

    char hello[256];
    snprintf(
        hello,
        sizeof(hello),
        "HELLO|CLIENT=ext-c-01|PROTO=RALF1|ROLE=%s|LASTSEQ=0",
        role
    );

    if (send_line(fd, hello) != 0) {
        perror("send HELLO");
        close(fd);
        return 1;
    }

    char line[4096];
    if (recv_line(fd, line, sizeof(line)) != 0) {
        perror("recv WELCOME");
        close(fd);
        return 1;
    }

    ralf_message_t welcome;
    if (ralf_parse_line(line, &welcome) != 0) {
        fprintf(stderr, "failed to parse WELCOME\n");
        close(fd);
        return 1;
    }

    printf("WELCOME type=%s GW=%s ROLE=%s\n",
           welcome.msg_type,
           ralf_get_field(&welcome, "GW") ? ralf_get_field(&welcome, "GW") : "?",
           ralf_get_field(&welcome, "ROLE") ? ralf_get_field(&welcome, "ROLE") : "?");

    /* Subscribe to all major channels for demonstration. */
    if (send_line(fd, "SUB|ROLE=CLEARING|CH=CLEARING|SYM=*") != 0 ||
        send_line(fd, "SUB|ROLE=DROP_COPY|CH=DROP_COPY|SYM=AAPL,MSFT") != 0 ||
        send_line(fd, "SUB|ROLE=AUDIT|CH=AUDIT|SYM=*") != 0) {
        perror("send SUB");
        close(fd);
        return 1;
    }

    printf("Subscribed as %s\n", role);

    while (recv_line(fd, line, sizeof(line)) == 0) {
        ralf_message_t msg;
        if (ralf_parse_line(line, &msg) != 0) {
            fprintf(stderr, "parse error: %s\n", line);
            continue;
        }
        printf("MSG type=%s CH=%s SEQ=%s SYM=%s\n",
               msg.msg_type,
               ralf_get_field(&msg, "CH") ? ralf_get_field(&msg, "CH") : "-",
               ralf_get_field(&msg, "SEQ") ? ralf_get_field(&msg, "SEQ") : "-",
               ralf_get_field(&msg, "SYM") ? ralf_get_field(&msg, "SYM") : "-");
    }

    close(fd);
    return 0;
}
